File: src/schema/application/system_resources.ts
import { bigint, integer, pgTable, text, doublePrecision } from 'drizzle-orm/pg-core';
import {
  getConfigDefaults,
  system_fields,
} from '@dna-platform/crdt-lww-postgres/build/schema/system';
import { primaryKey } from 'drizzle-orm/pg-core';
const config = (table) => ({
  pk: primaryKey({ columns: [table.id] }),
  ...getConfigDefaults.defaultIndexes('system_resources', table),
});
export const table = pgTable(
  'system_resources',
  {
    ...system_fields,
    id: text('id'), // Primary key
    // timestamp: timestamp('timestamp', { withTimezone: true }), // NOT NULL timestamp

    // CPU
    num_cpus: integer('num_cpus'),
    global_cpu_usage: doublePrecision('global_cpu_usage'),
    cpu_usages: text('cpu_usages'), // JSON string

    // RAM
    total_memory: bigint('total_memory', { mode: 'number' }),
    used_memory: bigint('used_memory', { mode: 'number' }),

    // Disk
    total_disk_space: bigint('total_disk_space', { mode: 'number' }),
    available_disk_space: bigint('available_disk_space', { mode: 'number' }),

    // I/O
    read_bytes: bigint('read_bytes', { mode: 'number' }),
    written_bytes: bigint('written_bytes', { mode: 'number' }),

    // Temperature
    temperatures: text('temperatures'), // JSON string
  },
  config,
);

File: bin/store/src/database/schema/tables/system_resources.rs
use crate::define_table_schema;
use crate::generated::schema::generator::diesel_schema_definition::{
    types::*, DieselTableDefinition,
};
use crate::{system_fields, system_foreign_keys, system_indexes};

/**
 * Auto-converted from TypeScript schema source.
 */
pub struct SystemResourcesTable;

define_table_schema! {
    hypertable: false,
    fields: {
        // System fields - common across all tables ( REQUIRED )
        system_fields!(),

        num_cpus: nullable(integer()),
        global_cpu_usage: nullable(doubleprecision()),
        cpu_usages: nullable(text()),
        total_memory: nullable(bigint()),
        used_memory: nullable(bigint()),
        total_disk_space: nullable(bigint()),
        available_disk_space: nullable(bigint()),
        read_bytes: nullable(bigint()),
        written_bytes: nullable(bigint()),
        temperatures: nullable(text()),
        device_id: nullable(text()),
    },
    indexes: {
        // System field indexes ( REQUIRED )
        system_indexes!("system_resources"),
    },
    foreign_keys: {
        // System field foreign keys ( REQUIRED )
        system_foreign_keys!("system_resources"),
        fk_system_resources_device_id: { columns: ["device_id"], foreign_table: "devices", foreign_columns: ["id"], on_delete: "no action", on_update: "no action" },
    }
}
